src/careamics/models/lvae/layers.py

```diff
@@ -1333,666 +1332,4 @@ def forward(
-        self,
-        p: torch.distributions.normal.Normal,
-        p_params: torch.Tensor,
-        q: torch.distributions.normal.Normal,
-        q_params: torch.Tensor,
-        mode_pred: bool,
-        analytical_kl: bool,
-        z: torch.Tensor,
-    ) -> Dict[str, torch.Tensor]:
-        """
-        Compute KL (analytical or MC estimate) and then process it, extracting composed versions of the metric.
-        Specifically, the different versions of the KL loss terms are:
-            - `kl_elementwise`: KL term for each single element of the latent tensor [Shape: (batch, ch, h, w)].
-            - `kl_samplewise`: KL term associated to each sample in the batch [Shape: (batch, )].
-            - `kl_samplewise_restricted`: KL term only associated to the portion of the latent tensor that is
-            used for prediction and summed over channel and spatial dimensions [Shape: (batch, )].
-            - `kl_channelwise`: KL term associated to each sample and each channel [Shape: (batch, ch, )].
-            - `kl_spatial`: KL term summed over the channels, i.e., retaining the spatial dimensions [Shape: (batch, h, w)]
-
-        Parameters
-        ----------
-        p: torch.distributions.normal.Normal
-            The prior generative distribution p(z_i|z_{i+1}) (or p(z_L)).
-        p_params: torch.Tensor
-            The parameters of the prior generative distribution.
-        q: torch.distributions.normal.Normal
-            The inference distribution q(z_i|z_{i+1}) (or q(z_L|x)).
-        q_params: torch.Tensor
-            The parameters of the inference distribution.
-        mode_pred: bool
-            Whether the model is in prediction mode.
-        analytical_kl: bool
-            Whether to compute the KL divergence analytically or using Monte Carlo estimation.
-        z: torch.Tensor
-            The sampled latent tensor.
-        """
-        kl_samplewise_restricted = None
-
-        if mode_pred is False:  # if not in prediction mode
-            # KL term for each single element of the latent tensor [Shape: (batch, ch, h, w)]
-            if analytical_kl:
-                kl_elementwise = kl_divergence(q, p)
-            else:
-                kl_elementwise = kl_normal_mc(z, p_params, q_params)
-
-            # KL term only associated to the portion of the latent tensor that is used for prediction and
-            # summed over channel and spatial dimensions. [Shape: (batch, )]
-            # NOTE: vanilla_latent_hw is the shape of the latent tensor used for prediction, hence
-            # the restriction has shape [Shape: (batch, ch, vanilla_latent_hw[0], vanilla_latent_hw[1])]
-            if self._restricted_kl:
-                pad = (kl_elementwise.shape[-1] - self._vanilla_latent_hw) // 2
-                assert pad > 0, "Disable restricted kl since there is no restriction."
-                tmp = kl_elementwise[..., pad:-pad, pad:-pad]
-                kl_samplewise_restricted = tmp.sum((1, 2, 3))
```

Comment:
Talking about what could not work in the 3D case if we restore this (as we should): all the aggregations of KL clearly need to be changed!

Reply:
Please change also the comments accordingly, so that they reflect the actual dimensions of tensors (otherwise it is a mess to understand).

Reply:
Good point!